ast/desugar/Desugar.cc

```diff
@@ -1534,6 +1544,13 @@ ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what)
                 values.reserve(hash->pairs.size()); // overapproximation in case there are KwSpats
                 ExpressionPtr lastMerge;
 
+                Send::ARGS_store mergeEntries;
+                mergeEntries.reserve(hash->pairs.size());
+
+                // Desguar
+                //   {**x, a: 'a', **y, remaining}
+                // into
+                //   <Magic>.<merge-hash>(<Magic>.<to-hash-dup>(y), {a: 'a'}, <Magic>.<to-hash-nodup>(x), remaining)
```

Comment:
It looks like the VM will turn this into something more like:

```
t = <Magic>.merge-into({}, x)
t = <Magic>.add-entries(t, :a, 'a')
t = <Magic>.merge-into(t, y)
# not entirely sure how `remaining` plays in here, so ignoring it
```

That seems slightly better, because we're not explicitly constructing a hash for `{a: 'a'}` and we don't need to worry about no-dup vs. dup because merge-into + the initial fresh hash takes care of that for us?

Reply:
Yep that's a good point. As long as we're threading new variables through as well, this should behave just like our existing merge intrinsics, so I think that I'll rework it to look like this instead.

Reply:
I've implemented this new strategy, and updated the description to indicate what the new desugaring looks like.